packages/python/pyfora/src/core/PyObjectPtr.hpp

```diff
@@ -0,0 +1,232 @@
+/***************************************************************************
+   Copyright 2016 Ufora Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+****************************************************************************/
+#pragma once
+
+#include <Python.h>
+
+#include <cstddef>
+
+
+class PyObjectPtr {
+public:
+    PyObjectPtr(PyObject* pyObject)
+        : mPtr(pyObject)
+        {
+        }
+
+    PyObjectPtr()
+        : mPtr(nullptr)
+        {
+        }
+
+    static PyObjectPtr Incremented(PyObject* pyObject)
+        {
+        PyObjectPtr tr = pyObject;
+        Py_XINCREF(pyObject);
+        return tr;
+        }
+
+    ~PyObjectPtr()
+        {
+        Py_XDECREF(mPtr);
+        }
+
+    PyObject* get() const
+        {
+        return mPtr;
+        }
+
+    PyObject* operator->() { return mPtr; }
+
+    PyObjectPtr(const PyObjectPtr& p)
+        : mPtr(p.mPtr)
+        {
+        Py_XINCREF(mPtr);
+        }
+
+    PyObjectPtr& operator=(const PyObjectPtr& p)
+        {
+        PyObject* const old = mPtr;
+        mPtr = p.mPtr;
+        Py_XINCREF(mPtr);
+        Py_XDECREF(old);
+        return *this;
+        }
+
```

Comment:
a lot of the things these PyObject*s get casted to are macros though! (your point still stands though)